addons/binding/org.openhab.binding.connectsdk/src/main/java/org/openhab/binding/connectsdk/handler/AbstractChannelHandler.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) 2010-2015, openHAB.org and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.connectsdk.handler;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.connectsdk.device.ConnectableDevice;
+import com.connectsdk.service.capability.listeners.ResponseListener;
+import com.connectsdk.service.command.ServiceCommandError;
+import com.connectsdk.service.command.ServiceSubscription;
+
+/**
+ * An abstract implementation of ChannelHander which serves as a base class for all concrete instances.
+ *
+ * @author Sebastian Prehn
+ * @since 1.8.0
+ */
+public abstract class AbstractChannelHandler<T> implements ChannelHandler {
+    private static final Logger logger = LoggerFactory.getLogger(AbstractChannelHandler.class);
+
+    // IP to Subscriptions map
+    private Map<String, ServiceSubscription<T>> subscriptions;
+
+    // lazy init
+    private synchronized Map<String, ServiceSubscription<T>> getSubscriptions() {
+        if (subscriptions == null) {
+            subscriptions = new ConcurrentHashMap<String, ServiceSubscription<T>>();
+        }
+        return subscriptions;
+    }
+
+    @Override
+    public void onDeviceReady(final ConnectableDevice device, final String channelId, final ConnectSDKHandler handler) {
+        // NOP
+    }
+
+    @Override
+    public void onDeviceRemoved(final ConnectableDevice device, final String channelId,
+            final ConnectSDKHandler handler) {
+        // NOP
+    }
+
+    @Override
+    public final synchronized void refreshSubscription(final ConnectableDevice device, final String channelId,
+            final ConnectSDKHandler handler) {
+        removeAnySubscription(device);
```

Comment:
This is because the capabilities of a connectsdk device change as it goes through its lifecycle. It is the safest to clear subscriptions and add them again.

Reply:
OK, I understand it now.